File: leetcode-236.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        self.ans = root
        def dfs(u):
            if u is None:
                return False
            cur = u ==p or u == q
            le = dfs(u.left)
            re = dfs(u.right)
            if (le and re ) or (cur and (le or re)):
                self.ans = u
            return le or re or cur
        dfs(root)
        return self.ans

File: leetcode/leetcode-236.cpp
/*
 * @File Name          :leetcode-236.cpp
 * @Author             :cc
 * @Version            :1.0.0
 * @Date               :2022/1/27 10:15
 * @Description        :
 * @Function List      :
 * @History            :
 */

#include "algorithm"

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
public:
    TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q) {
        if (root == nullptr) return nullptr;
        if (root == p || root == q) return root;
        TreeNode *l = lowestCommonAncestor(root->left, p, q);
        TreeNode *r = lowestCommonAncestor(root->right, p, q);
        if (l && r) return root;
        if (l) return l;
        if (r) return r;
        return nullptr;
    }
};